### ISSUE TITLE
Predictions with DIABLO: failure if old and new data set don't have the same set of markers

### ISSUE THREAD

**Issue by lbeltrame** [mixOmicsTeam/mixOmics#40]
If the data generated by DIABLO is tested then, with `predict` on a different data set that does not have the same genes and / or other markers like miRNA (an example is TCGA which uses very old miRNA annotations from Mirbase), the prediction fails with a rather cryptic error.

A look at the code shows that this is done by purpose:

```r
stop("Each 'newdata[[i]]' must include all the variables of 'object$X[[i]]'")
```

However this can be problematic when one wants to compare data obtained from different platforms: for example we have predictions made with DIABLO on gene expression and miRNA microarrays built later than what TCGA has, and thus testing the DIABLO prediction in TCGA itself is not possible.

**Comment by singha53-zz** [comment]
@lbeltrame Do you mean the features between training and test are different or that mapping of features is different between microarrays and TCGA (RNA-Seq data)?

The former applies to any machine learning algorithm, ie. the specific features measured must be the same between train and test, whereas the latter can be solved by mapping platform specific ids to feature symbols (e.g. gene symbols, miRNA, protein names). 

biomaRt is useful when mapping between different ids (e.g. Ensembl, RefSeq, gene symbol):
https://bioconductor.org/packages/release/bioc/html/biomaRt.html

**Comment by lbeltrame** [comment]
> @lbeltrame Do you mean the features between training and test are different or that mapping of features is different between microarrays and TCGA (RNA-Seq data)?

Yes, we're using in-house data (Agilent 8x60K microarrays) and we'd like to compare with the TCGA (using, at this point, the combined microarray platforms they used).

**Comment by singha53-zz** [comment]
Seems like there are two problems: A) mapping B) mismatched training and test sets due to differences in platforms.

A) Mapping:
1. convert your agilent ids to gene symbols: https://support.bioconductor.org/p/90261/
2. do the same with the TCGA microarrays

Result: 
- you will have datasets with the same feature identifier (ie. gene symbol)
- some ids may map to the same gene symbol, so average those to get a single feature
- some ids may map to nothing, so remove those
- the number of resulting gene symbols in each dataset may differ: e.g training (agilent), test (hg133, affymetrix gene chips etc, illumina etc).

B) mismatched training and test sets due to differences in platforms.
since your training and test data are different it would be naive to build on the training and evaluate the performance on the test without any preprocessing of the data. Some mitigation strategies include:

1. combine all data and correct for batch (platform)
this [paper](https://www-ncbi-nlm-nih-gov.ezproxy.library.ubc.ca/pmc/articles/PMC4499117/) uses ComBat to correct for platform differences and even gets around the different features present in the discovery and validation sets by making many discovery and validation sets based on overlapping features (see Figure 1).

2. randomly shuffle samples so that the training and test set each have similar numbers of platform specific sample data
this [paper](https://www-ncbi-nlm-nih-gov.ezproxy.library.ubc.ca/pmc/articles/PMC5697344/) mixes both microarray and rna-seq samples and re-creates the training and test set.

3. split your validation data into a dev and test set.
the deep learning literature suggests to set up a dev (development) set which is a partition of your validation set to optimize the training algorithm to (see [video](https://www.coursera.org/lecture/machine-learning-projects/train-dev-test-distributions-78P8f))

**Comment by lbeltrame** [comment]
Thanks a bunch for all the suggestions! We'll follow up with those.

**Comment by aljabadi** [comment]
Please feel free to follow up on this or similar analysis discussions on our [Discourse forum](https://mixomics-users.discourse.group/).
